idunn/datasources/hove/models.py

```diff
@@ -0,0 +1,338 @@
+            (-90, "left"),
+            (-45, "slight left"),
+            (0, "straight"),
+            (45, "slight right"),
+            (90, "right"),
+            (135, "sharp right"),
+            (180, "uturn"),
+        ]
+
+        _, closest = min(
+            (
+                min(abs(self.direction - angle), abs(360 - self.direction - angle)),
+                modifier,
+            )
+            for angle, modifier in mapping
+        )
+
+        return closest
+
+
+class DisplayInformations(BaseModel):
+    code: Optional[str]
+    label: Optional[str]
+    color: Optional[str]
+    text_color: Optional[str]
+    network: Optional[str]
+    direction: Optional[str]
+    physical_mode: Optional[str]
+
+    def as_api_transport_info(self) -> api.TransportInfo:
+        return api.TransportInfo(
+            num=self.code or self.label,
+            direction=self.direction,
+            lineColor=self.color,
+            lineTextColor=self.text_color,
+            network=self.network,
+        )
+
+
+class StopPoint(BaseModel):
+    name: str
+    coord: Coordinate
+
+
+class StopDateTime(BaseModel):
+    stop_point: StopPoint
+
+
+class SectionType(Enum):
+    PUBLIC_TRANSPORT = "public_transport"
+    STREET_NETWORK = "street_network"
+    WAITING = "waiting"
+    STAY_IN = "stay_in"
+    TRANSFER = "transfer"
+    CROW_FLY = "crow_fly"
+    ON_DEMAND_TRANSPORT = "on_demand_transport"
+    BSS_RENT = "bss_rent"
+    BSS_PUT_BACK = "bss_put_back"
+    BOARDING = "boarding"
+    LANDING = "landing"
+    ALIGHTING = "alighting"
+    PARK = "park"
+    RIDESHARING = "ridesharing"
+
+
+class Section(BaseModel):
+    sec_type: SectionType = Field(..., alias="type")
+    mode: Optional[IdunnTransportMode]
+    duration: int
+    path: List[Instruction] = []
+    # NOTE: this is always set when not waiting
+    # NOTE: this is not a purely valid geojson because of the properties field which is present in
+    #       a LineString geometry and which contains a list of objects instead of an object
+    geojson: Optional[dict]
+    display_informations: Optional[DisplayInformations]
+    stop_date_times: List[StopDateTime] = []
+
+    @validator("mode", pre=True)
+    def mode_from_navitia(cls, v):
+        return IdunnTransportMode.parse(v)
+
+    def cut_linestring(self) -> Iterator[Tuple[Instruction, Geometry]]:
+        """
+        Split path geometry at each instruction and wrap them together
+        """
+        all_coords = self.geojson["coordinates"]
+        first_pass = True
+        last_inst = None
+
+        for inst in self.path:
+            if not inst.instruction_start_coordinate:
+                continue
+
+            end_coord = inst.instruction_start_coordinate.to_position()
+
+            try:
+                end_index = next(
+                    i
+                    for i, coord in enumerate(all_coords)
+                    if abs(coord[0] - end_coord[0]) < 1e-6 and abs(coord[1] - end_coord[1]) < 1e-6
```

Comment:
1e-6 is something like 1meter ?

Reply:
Oh your right, that should be a documented constant, I put it here during the POC and forgot it after that.

Worst case scenario, for the latitude a variation of $1\mathrm{e}{-6}°$ leads to a variation of $$\frac{1\mathrm{e}{-6}}{180} \cdot 40 \cdot 10^{6} \approx 0.22m$$

This is mainly intended to deal with rounding errors, but 22cm seems like a fair margin :shrug: